packages/core/src/ui/UIRenderer.ts

```diff
@@ -0,0 +1,193 @@
+    if (onlyMVP) {
+      this._updateProjectionRelatedShaderData(context, worldMatrix, true);
+    } else {
+      this._updateWorldViewRelatedShaderData(context, worldMatrix, true);
+    }
+  }
+
+  /**
+   * @internal
+   */
+  override _prepareRender(context: RenderContext): void {
+    // Update once per frame per renderer, not influenced by batched
+    if (this._renderFrameCount !== this.engine.time.frameCount) {
+      this._update(context);
+    }
+
+    this._render(context);
+
+    // union camera global macro and renderer macro.
+    ShaderMacroCollection.unionCollection(
+      context.camera._globalShaderMacro,
+      this.shaderData._macroCollection,
+      this._globalShaderMacro
+    );
+  }
+
+  /**
+   * @internal
+   */
+  override _onEnableInScene(): void {
+    const componentsManager = this.scene._componentsManager;
+    if (this._overrideUpdate) {
+      componentsManager.addOnUpdateRenderers(this);
+    }
+  }
+
+  /**
+   * @internal
+   */
+  override _onDisableInScene(): void {
+    const componentsManager = this.scene._componentsManager;
+    if (this._overrideUpdate) {
+      componentsManager.removeOnUpdateRenderers(this);
+    }
+  }
+
+  /**
+   * @internal
+   */
+  _setGroup(group: CanvasGroup): void {
+    this._group = group;
+    const alpha = group ? group._globalAlpha : 1;
+    if (this._alpha !== alpha) {
+      this._alpha = alpha;
+      this._dirtyUpdateFlag |= UIRendererUpdateFlags.Alpha;
+    }
+  }
+
+  /**
+   * @internal
+   */
+  _getChunkManager(): PrimitiveChunkManager {
+    return this.engine._batcherManager.primitiveChunkManagerUI;
+  }
+
+  /** @internal */
+  _raycast(ray: Ray, out: HitResult, distance: number = Number.MAX_SAFE_INTEGER): boolean {
+    const entity = this._entity;
+    const plane = UIRenderer._tempPlane;
+    const transform = entity.transform;
+    const normal = plane.normal.copyFrom(transform.worldForward);
+    plane.distance = -Vector3.dot(normal, transform.worldPosition);
+    ray.intersectPlane(plane);
+    const curDistance = ray.intersectPlane(plane);
+    if (curDistance >= 0 && curDistance < distance) {
+      const hitPointWorld = ray.getPoint(curDistance, UIRenderer._tempVec30);
+      const worldMatrixInv = UIRenderer._tempMat;
+      Matrix.invert(this.entity.transform.worldMatrix, worldMatrixInv);
+      const hitPointLocal = UIRenderer._tempVec31;
+      Vector3.transformCoordinate(hitPointWorld, worldMatrixInv, hitPointLocal);
+      if (this.isRaycastLocationValid(hitPointLocal)) {
+        out.distance = curDistance;
+        out.entity = entity;
+        out.normal.copyFrom(normal);
+        out.point.copyFrom(hitPointWorld);
+        return true;
+      }
+    }
+    return false;
+  }
+
+  protected isRaycastLocationValid(hitPoint: Vector3): boolean {
+    const { x, y } = hitPoint;
+    const uiTransform = <UITransform>this._transform;
+    const { x: width, y: height } = uiTransform.size;
+    const { x: pivotX, y: pivotY } = uiTransform.pivot;
```

Comment:
Above two lines can be optimize like this:
const { x, y } = hitPoint;
const { size: { x: width, y: height }, pivot: { x: pivotX, y: pivotY } } = <UITransform>this._transform;